burn-tensor/src/tensor/ops/int_tensor.rs

```diff
@@ -458,6 +458,64 @@ pub trait IntTensorOps<B: Backend> {
     /// The result of the addition.
     fn int_add_scalar<const D: usize>(lhs: IntTensor<B, D>, rhs: IntElem<B>) -> IntTensor<B, D>;
 
+    /// Elementwise power.
+    ///
+    /// # Arguments
+    ///
+    /// * `lhs` - The left hand side tensor.
+    /// * `rhs` - The right hand side tensor.
+    ///
+    /// # Returns
+    ///
+    /// The elements of `lhs` raised to the power of the elements of `rhs`.
+    fn int_pow<const D: usize>(lhs: IntTensor<B, D>, rhs: IntTensor<B, D>) -> IntTensor<B, D>;
+
+    /// Elementwise power with a scalar.
+    ///
+    /// # Arguments
+    ///
+    /// * `lhs` - The left hand side tensor.
+    /// * `rhs` - The right hand side scalar.
+    ///
+    /// # Returns
+    ///
+    /// The elements of `lhs` raised to the value of `rhs`.
+    fn int_pow_scalar<const D: usize>(lhs: IntTensor<B, D>, rhs: IntElem<B>) -> IntTensor<B, D>;
+
+    /// Elementwise power with a floatTensor.
+    ///
+    /// # Arguments
+    ///
+    /// * `lhs` - The left hand side tensor.
+    /// * `rhs` - The right hand side floatTensor.
+    ///
+    /// # Returns
+    ///
+    /// The elements of `lhs` raised to the value of `rhs`. Result is an IntTensor.
+    fn int_pow_float<const D: usize>(
+        lhs: IntTensor<B, D>,
+        rhs: FloatTensor<B, D>,
+    ) -> IntTensor<B, D> {
+        Self::int_pow(lhs, B::into_int(rhs))
+    }
```

Comment:
It can cause precision issues. I think we should only support `pow` and `powf` where the `rhs` is always `f32`. So in the case of an int, we would have `int_pow` and `int_powf`, and in the case of a float, we would have `pow` and `powf`. `pow` is always an elementwise operation, and `powf` is always a scalar operation. What are your thoughts?

Reply:
sounds great, it actually makes it a bit easier to implement the backends, and that brings down the number of functions to implement from 4 to 3 per backend(powf already exist)

Reply:
Hmm, maybe to be cleaner we might have all 4 variants:

- `powi` (Rhs = Int Tensor)
- `powf` (Rhs = Float Tensor)
- `powi_scalar` (Rhs = Int Scalar)
- `powf_scalar` (Rhs = Float Scalar)

We could provide default implementations for `powi` and `powi_scalar` where `powi => powf` and `powi_scalar => powf_scalar`. Using the float implementation by default is necessary if we want to keep the precision.

Reply:
> Hmm, maybe to be cleaner we might have all 4 variants:
> 
> * `powi` (Rhs = Int Tensor)
> * `powf` (Rhs = Float Tensor)
> * `powi_scalar` (Rhs = Int Scalar)
> * `powf_scalar` (Rhs = Float Scalar)
> 
> We could provide default implementations for `powi` and `powi_scalar` where `powi => powf` and `powi_scalar => powf_scalar`. Using the float implementation by default is necessary if we want to keep the precision.

Sorry just saw this, the main problem is `powf` predates this PR, it already exist in burn. I can change the naming scheme, but that will break any downstream projects that are calling that function

Reply:
> Hmm, maybe to be cleaner we might have all 4 variants:
> 
> * `powi` (Rhs = Int Tensor)
> * `powf` (Rhs = Float Tensor)
> * `powi_scalar` (Rhs = Int Scalar)
> * `powf_scalar` (Rhs = Float Scalar)
> 
> We could provide default implementations for `powi` and `powi_scalar` where `powi => powf` and `powi_scalar => powf_scalar`. Using the float implementation by default is necessary if we want to keep the precision.

A) do you want me to change the name (and maybe the signature) of the existing `powf` function to make it fit the convention
B) am I correct in understanding that in total we'll have 8 functions (`[int_]pow{i,f}[_scalar]`) but all but 2 will have default implementations like the below:
```rust
fn int_powi<const D: usize>(lhs: IntTensor<B, D>, rhs: IntTensor<B, D>) -> IntTensor<B, D> {
        B::into_int(B::powf(B::int_into_float(lhs), B::int_into_float(rhs)))
    }
```

Reply:
A) Yes I prefer having a consistent API than avoid a simple breaking change at this stage.
B) Almost, I think we should only provide a default implementation for 4 variants:

```rust
fn int_powi<const D: usize>(lhs: IntTensor<B, D>, rhs: IntTensor<B, D>) -> IntTensor<B, D> {
        B::int_powf(lhs, B::int_into_float(rhs)))
}
fn int_powi_scalar<const D: usize>(lhs: IntTensor<B, D>, rhs: IntElem<B>) -> IntTensor<B, D> {
        B::int_powf_scalar(lhs,  rhs.elem()))
}
fn powi<const D: usize>(lhs: FloatTensor<B, D>, rhs: IntTensor<B, D>) -> FloatTensor<B, D> {
        B::powf(lhs, B::int_into_float(rhs)))
}
fn powi_scalar<const D: usize>(lhs: FloatTensor<B, D>, rhs: IntElemr<B, D>) -> FloatTensor<B, D> {
        B::int_powf_scalar(lhs,  rhs.elem()))
}
 ```

Reply:
just saw this, I'll implement these changes in the next commit, might be tomorrow though

Reply:
> A) Yes I prefer having a consistent API than avoid a simple breaking change at this stage. B) Almost, I think we should only provide a default implementation for 4 variants:
> 
> ```rust
> fn int_powi<const D: usize>(lhs: IntTensor<B, D>, rhs: IntTensor<B, D>) -> IntTensor<B, D> {
>         B::int_powf(lhs, B::int_into_float(rhs)))
> }
> fn int_powi_scalar<const D: usize>(lhs: IntTensor<B, D>, rhs: IntElem<B>) -> IntTensor<B, D> {
>         B::int_powf_scalar(lhs,  rhs.elem()))
> }
> fn powi<const D: usize>(lhs: FloatTensor<B, D>, rhs: IntTensor<B, D>) -> FloatTensor<B, D> {
>         B::powf(lhs, B::int_into_float(rhs)))
> }
> fn powi_scalar<const D: usize>(lhs: FloatTensor<B, D>, rhs: IntElemr<B, D>) -> FloatTensor<B, D> {
>         B::int_powf_scalar(lhs,  rhs.elem()))
> }
> ```

won't the int_powf still need to call the regular `tensor::powf`? It does involve two conversions RHS to tensor, and output to Int, but if `powi` is just calling `powf`, then those conversions will be necessary unless the backend has a `mapv` iterator like ndarray

Reply:
the original request has been resolved.